java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -0,0 +1,300 @@
+    this.length = length;
+  }
+
+  public boolean visit(BaseFixedWidthVector left) {
+    return compareBaseFixedWidthVectors(left);
+  }
+
+  public boolean visit(BaseVariableWidthVector left) {
+    return compareBaseVariableWidthVectors(left);
+  }
+
+  public boolean visit(ListVector left) {
+    return compareListVectors(left);
+  }
+
+  public boolean visit(FixedSizeListVector left) {
+    return compareFixedSizeListVectors(left);
+  }
+
+  public boolean visit(NonNullableStructVector left) {
+    return compareStructVectors(left);
+  }
+
+  public boolean visit(UnionVector left) {
+    return compareUnionVectors(left);
+  }
+
+  public boolean visit(ZeroVector left) {
+    return false;
+  }
+
+  protected boolean compareUnionVectors(UnionVector left) {
+    if (!left.getField().getType().equals(right.getField().getType())) {
+      return false;
+    }
+
+    UnionVector rightVector = (UnionVector) right;
+
+    List<FieldVector> leftChildrens = left.getChildrenFromFields();
+    List<FieldVector> rightChildrens = rightVector.getChildrenFromFields();
+
+    if (leftChildrens.size() != rightChildrens.size()) {
+      return false;
+    }
+
+    for (int k = 0; k < leftChildrens.size(); k++) {
+      RangeEqualsVisitor visitor = new RangeEqualsVisitor(rightChildrens.get(k),
+          leftStart, rightStart, length);
+      if (!leftChildrens.get(k).accept(visitor)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected boolean compareStructVectors(NonNullableStructVector left) {
+    if (!left.getField().getType().equals(right.getField().getType())) {
+      return false;
+    }
+
+    NonNullableStructVector rightVector = (NonNullableStructVector) right;
+
+    if (!left.getChildFieldNames().equals(rightVector.getChildFieldNames())) {
+      return false;
+    }
+
+    List<ValueVector> leftChildrens = new ArrayList<>();
+    List<ValueVector> rightChildrens = new ArrayList<>();
+
+    for (String child : left.getChildFieldNames()) {
+      ValueVector v = left.getChild(child);
+      if (v != null) {
+        leftChildrens.add(v);
+      }
+    }
+
+    for (String child : rightVector.getChildFieldNames()) {
+      ValueVector v = rightVector.getChild(child);
+      if (v != null) {
+        rightChildrens.add(v);
+      }
+    }
+
+    if (leftChildrens.size() != rightChildrens.size()) {
+      return false;
+    }
+
+    for (int k = 0; k < leftChildrens.size(); k++) {
+      RangeEqualsVisitor visitor = new RangeEqualsVisitor(rightChildrens.get(k),
+          leftStart, rightStart, length);
+      if (!leftChildrens.get(k).accept(visitor)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected boolean compareBaseFixedWidthVectors(BaseFixedWidthVector left) {
+
+    if (!left.getField().getType().equals(right.getField().getType())) {
+      return false;
+    }
+
+    int typeWidth = left.getTypeWidth();
+
+    for (int i = 0; i < length; i++) {
+      int leftIndex = leftStart + i;
```

Comment:
we can have fast paths for all values being non-null, but they can be addressed in follow up patches..

Reply:
Sure, thanks for your comments :)